### PR TITLE
JDK-8302040: Port fdlibm sqrt to Java

### PR DESCRIPTION
**Pull request by jddarcy**
The wheel of FDLIBM porting turns a notch and sqrt comes into play.

While the sqrt operation usually has a hardware implementation that is intrinsified, for completeness a software implementation should be available as well.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8302040](https://bugs.openjdk.org/browse/JDK-8302040): Port fdlibm sqrt to Java

### Reviewers
 * [Brian Burkhalter](https://openjdk.org/census#bpb) (@bplb - **Reviewer**) ⚠️ Review applies to [7d39761c](https://git.openjdk.org/jdk/pull/12736/files/7d39761c011be09337ba5c87e81f264ddc9c7cf1)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Andrey Turbanov](https://openjdk.org/census#aturbanov) (@turbanoff - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12736/head:pull/12736` \
`$ git checkout pull/12736`

Update a local copy of the PR: \
`$ git checkout pull/12736` \
`$ git pull https://git.openjdk.org/jdk pull/12736/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12736`

View PR using the GUI difftool: \
`$ git pr show -t 12736`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12736.diff">https://git.openjdk.org/jdk/pull/12736.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/lang/StrictMath.java [hunk 1]
```diff
@@ -309,8 +309,10 @@ public static double log10(double a) {
      * @param   a   a value.
      * @return  the positive square root of {@code a}.
      */
-    @IntrinsicCandidate
-    public static native double sqrt(double a);
+    // @IntrinsicCandidate -- Temporarily disable for testing
```
Comment (jddarcy) [inline src/java.base/share/classes/java/lang/StrictMath.java:312]: I'll revert commenting out the intrinsic annotation before pushing.

**Comment by jddarcy** [comment]
In local testing, both the transliteration port and the Fdlibm.java port are consistent with hardware sqrt when run against all float values.

Comparing the various ports, original C sources vs transliteration port:

```
$ diff -w Sqrt.c  Sqrt.translit.java 
1c1
< /* __ieee754_sqrt(x)
---
>     /**
69a70,71
>     static class Sqrt {
>         private static final double    one     = 1.0, tiny=1.0e-300;
71,86c73,74
< #include "fdlibm.h"
< 
< #ifdef __STDC__
< static  const double    one     = 1.0, tiny=1.0e-300;
< #else
< static  double  one     = 1.0, tiny=1.0e-300;
< #endif
< 
< #ifdef __STDC__
<         double __ieee754_sqrt(double x)
< #else
<         double __ieee754_sqrt(x)
<         double x;
< #endif
< {
<         double z;
---
>         public static double compute(double x) {
>             double z = 0.0;
88c76
<         unsigned r,t1,s1,ix1,q1;
---
>             /*unsigned*/ int r,t1,s1,ix1,q1;
110c98
<                 ix0 |= (ix1>>11); ix1 <<= 21;
---
>                     ix0 |= (ix1>>>11); ix1 <<= 21; // unsigned shift
114c102
<             ix0 |= (ix1>>(32-i));
---
>                 ix0 |= (ix1>>>(32-i)); // unsigned shift
119,120c107,108
<         if(m&1){        /* odd m, double x to make it even */
<             ix0 += ix0 + ((ix1&sign)>>31);
---
>             if((m&1) != 0){        /* odd m, double x to make it even */
>                 ix0 += ix0 + ((ix1&sign)>>>31); // unsigned shift
126c114
<         ix0 += ix0 + ((ix1&sign)>>31);
---
>             ix0 += ix0 + ((ix1&sign)>>>31); // unsigned shift
138c126
<             ix0 += ix0 + ((ix1&sign)>>31);
---
>                 ix0 += ix0 + ((ix1&sign)>>>31); // unsigned shift
140c128
<             r>>=1;
---
>                 r>>>=1; // unsigned shift
147c135
<             if((t<ix0)||((t==ix0)&&(t1<=ix1))) {
---
>                 if((t<ix0)||((t==ix0)&&(Integer.compareUnsigned(t1, ix1) <= 0 ))) { // t1<=ix1
151c139
<                 if (ix1 < t1) ix0 -= 1;
---
>                     if (Integer.compareUnsigned(ix1, t1) < 0) ix0 -= 1; // ix1 < t1
155c143
<             ix0 += ix0 + ((ix1&sign)>>31);
---
>                 ix0 += ix0 + ((ix1&sign)>>>31);
157c145
<             r>>=1;
---
>                 r>>>=1; // unsigned shift
165c153
<                 if (q1==(unsigned)0xffffffff) { q1=0; q += 1;}
---
>                     if (q1==0xffffffff) { q1=0; q += 1;}
167c155
<                     if (q1==(unsigned)0xfffffffe) q+=1;
---
>                         if (q1==0xfffffffe) q+=1;
174c162
<         ix1 =  q1>>1;
---
>             ix1 =  q1>>>1; // unsigned shift
177,178c165,168
<         __HI(z) = ix0;
<         __LO(z) = ix1;
---
>             // __HI(z) = ix0;
>             z = __HI(z, ix0);
>             // __LO(z) = ix1;
>             z = __LO(z, ix1);
180a171
>     }

```

The the variables declared unsigned, checked to use unsigned right shifts (>>>) and to use the unsigned comparison methods.

```
$ diff -w  Sqrt.translit.java Sqrt.fdlibm.java 
71c71
<         private static final double    one     = 1.0, tiny=1.0e-300;
---
>         private Sqrt() {throw new UnsupportedOperationException();}
73c73,75
<         public static double compute(double x) {
---
>         private static final double tiny = 1.0e-300;
> 
>         static double compute(double x) {
75c77
<             int     sign = (int)0x80000000;
---
>             int sign = 0x8000_0000;
79,80c81,82
<             ix0 = __HI(x);                  /* high word of x */
<             ix1 = __LO(x);          /* low word of x */
---
>             ix0 = __HI(x);  // high word of x
>             ix1 = __LO(x);  // low word of x
82,85c84,86
<             /* take care of Inf and NaN */
<             if((ix0&0x7ff00000)==0x7ff00000) {
<                 return x*x+x;               /* sqrt(NaN)=NaN, sqrt(+inf)=+inf
<                                                sqrt(-inf)=sNaN */
---
>             // take care of Inf and NaN
>             if((ix0 & 0x7ff0_0000) == 0x7ff0_0000) {
>                 return x*x + x; // sqrt(NaN)=NaN, sqrt(+inf)=+inf, sqrt(-inf)=sNaN
87c88
<             /* take care of zero */
---
>             // take care of zero
89c90,91
<                 if(((ix0&(~sign))|ix1)==0) return x;/* sqrt(+-0) = +-0 */
---
>                 if (((ix0 & (~sign)) | ix1) == 0)
>                     return x; // sqrt(+-0) = +-0
91c93
<                     return (x-x)/(x-x);             /* sqrt(-ve) = sNaN */
---
>                     return (x-x)/(x-x); // sqrt(-ve) = sNaN
93c95
<             /* normalize x */
---
>             // normalize x
95c97
<             if(m==0) {                              /* subnormal x */
---
>             if (m == 0) { // subnormal x
98c100,104
<                     ix0 |= (ix1>>>11); ix1 <<= 21; // unsigned shift
---
>                     ix0 |= (ix1 >>> 11); // unsigned shift
>                     ix1 <<= 21;
>                 }
>                 for(i = 0; (ix0 & 0x0010_0000) == 0; i++) {
>                     ix0 <<= 1;
100d105
<                 for(i=0;(ix0&0x00100000)==0;i++) ix0<<=1;
105,107c110,112
<             m -= 1023;      /* unbias exponent */
<             ix0 = (ix0&0x000fffff)|0x00100000;
<             if((m&1) != 0){        /* odd m, double x to make it even */
---
>             m -= 1023;      // unbias exponent */
>             ix0 = (ix0 & 0x000f_ffff) | 0x0010_0000;
>             if ((m & 1) != 0){        // odd m, double x to make it even
111c116
<             m >>= 1;        /* m = [m/2] */
---
>             m >>= 1;        // m = [m/2]
113c118
<             /* generate sqrt(x) bit by bit */
---
>             // generate sqrt(x) bit by bit
116,117c121,122
<             q = q1 = s0 = s1 = 0;   /* [q,q1] = sqrt(x) */
<             r = 0x00200000;         /* r = moving bit from right to left */
---
>             q = q1 = s0 = s1 = 0;   // [q,q1] = sqrt(x)
>             r = 0x0020_0000;        // r = moving bit from right to left
135c140,141
<                 if((t<ix0)||((t==ix0)&&(Integer.compareUnsigned(t1, ix1) <= 0 ))) { // t1<=ix1
---
>                 if((t < ix0) ||
>                    ((t == ix0) && (Integer.compareUnsigned(t1, ix1) <= 0 ))) { // t1 <= ix1
137c143,145
<                     if(((t1&sign)==sign)&&(s1&sign)==0) s0 += 1;
---
>                     if (((t1 & sign) == sign) && (s1 & sign) == 0) {
>                         s0 += 1;
>                     }
139c147,149
<                     if (Integer.compareUnsigned(ix1, t1) < 0) ix0 -= 1; // ix1 < t1
---
>                     if (Integer.compareUnsigned(ix1, t1) < 0) {  // ix1 < t1
>                         ix0 -= 1;
>                     }
148c158
<             /* use floating add to find out rounding direction */
---
>             // use floating add to find out rounding direction
150,155c160,169
<                 z = one-tiny; /* trigger inexact flag */
<                 if (z>=one) {
<                     z = one+tiny;
<                     if (q1==0xffffffff) { q1=0; q += 1;}
<                     else if (z>one) {
<                         if (q1==0xfffffffe) q+=1;
---
>                 z = 1.0 - tiny; // trigger inexact flag
>                 if (z >= 1.0) {
>                     z = 1.0 + tiny;
>                     if (q1 == 0xffff_ffff) {
>                         q1 = 0;
>                         q += 1;
>                     } else if (z > 1.0) {
>                         if (q1 == 0xffff_fffe) {
>                             q += 1;
>                         }
157c171
<                     } else
---
>                     } else {
161c175,176
<             ix0 = (q>>1)+0x3fe00000;
---
>             }
>             ix0 = (q >> 1) + 0x3fe0_0000;
163c178,180
<             if ((q&1)==1) ix1 |= sign;
---
>             if ((q & 1) == 1) {
>                 ix1 |= sign;
>             }
165d181
<             // __HI(z) = ix0;
167d182
<             // __LO(z) = ix1;
```

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 2]
```diff
@@ -504,6 +504,451 @@ static double compute(double y, double x) {
...
+                }
+            }
+            ix0 = (q >> 1) + 0x3fe0_0000;
+            ix1 =  q1 >>> 1; // unsigned shift
+            if ((q & 1) == 1) {
+                ix1 |= sign;
+            }
+            ix0 += (m << 20);
+            z = __HI(z, ix0);
+            z = __LO(z, ix1);
```
Comment (jddarcy) [inline src/java.base/share/classes/java/lang/FdLibm.java:689]: I was tempted to rewrite these line as something like:
`return Double.longBitsToDouble((long)ix0 << 32 | (long)ix1);`
and may do so in a future refactoring.

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 3]
```diff
@@ -504,6 +504,451 @@ static double compute(double y, double x) {
...
+            return z;
+        }
+    }
+
+    // The following comment is supplementary information from the FDLIBM sources.
+
+    /*
```
Comment (jddarcy) [inline src/java.base/share/classes/java/lang/FdLibm.java:696]: I didn't think it was necessary to preserve this supplemental information in both the transliteration port and the mainline sources.

File: test/jdk/java/lang/StrictMath/FdlibmTranslit.java [hunk 4]
```diff
@@ -504,6 +508,178 @@ static double compute(double y, double x) {
...
+                t  = s0;
+                if((t<ix0)||((t==ix0)&&(Integer.compareUnsigned(t1, ix1) <= 0 ))) { // t1<=ix1
+                    s1  = t1+r;
+                    if(((t1&sign)==sign)&&(s1&sign)==0) s0 += 1;
+                    ix0 -= t;
+                    if (Integer.compareUnsigned(ix1, t1) < 0) ix0 -= 1; // ix1 < t1
+                    ix1 -= t1;
+                    q1  += r;
+                }
+                ix0 += ix0 + ((ix1&sign)>>>31);
```
Comment (rgiulietti) [inline test/jdk/java/lang/StrictMath/FdlibmTranslit.java:653]: You may want to add the traditional `// unsigned shift` comment even here, just to be consistent.

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 5]
```diff
@@ -504,6 +504,451 @@ static double compute(double y, double x) {
...
+                        s0 += 1;
+                    }
+                    ix0 -= t;
+                    if (Integer.compareUnsigned(ix1, t1) < 0) {  // ix1 < t1
+                        ix0 -= 1;
+                    }
+                    ix1 -= t1;
+                    q1  += r;
+                }
+                ix0 += ix0 + ((ix1 & sign) >>> 31);
```
Comment (rgiulietti) [inline src/java.base/share/classes/java/lang/FdLibm.java:659]: You may want to add the traditional `// unsigned shift` comment even here, just to be consistent.

File: src/java.base/share/classes/java/lang/FdLibm.java (hunk 2 above)
Comment (rgiulietti) [reply src/java.base/share/classes/java/lang/FdLibm.java:689]: That should read
`return Double.longBitsToDouble((long)ix0 << 32 | ix1 & 0xffff_ffffL);`

File: test/jdk/java/lang/StrictMath/ExhaustingTests.java [hunk 6]
```diff
@@ -66,7 +66,10 @@ public static void main(String... args) {
     private static long testUnaryMethods() {
         long failures = 0;
         UnaryTestCase[] testCases = {
-         // new UnaryTestCase("sqrt",  FdlibmTranslit::sqrt,  StrictMath::sqrt,  DEFAULT_SHIFT),
+            // Since sqrt is correctly rounded, could additionaly the
+            // transliteration or StrictMath sqrt methods to against
+            // Math::sqrt.
```
Comment (rgiulietti) [inline test/jdk/java/lang/StrictMath/ExhaustingTests.java:71]: I guess this comment has been messed up by some copy&paste activities?
Or is it a ChatGPT output?

**Comment by rgiulietti** [comment]
Otherwise LGTM

File: test/jdk/java/lang/StrictMath/ExhaustingTests.java (hunk 6 above)
Comment (jddarcy) [reply test/jdk/java/lang/StrictMath/ExhaustingTests.java]: Fixed mangled text.

File: src/java.base/share/classes/java/lang/FdLibm.java (hunk 2 above)
Comment (jddarcy) [reply src/java.base/share/classes/java/lang/FdLibm.java]: Added new method to combine high and low. In a future refactoring pass, I'll use the method in other locations in FDLIBM where it is applicable.

**Review by bplb — APPROVED** [review]
Go for it.

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 7]
```diff
@@ -504,6 +513,449 @@ static double compute(double y, double x) {
...
+                r >>>= 1; // unsigned shift
+            }
+
+            r = sign;
+            while (r != 0) {
+                t1 = s1 + r;
+                t  = s0;
+                if((t < ix0) ||
+                   ((t == ix0) && (Integer.compareUnsigned(t1, ix1) <= 0 ))) { // t1 <= ix1
+                    s1  = t1 + r;
```
Comment (turbanoff) [inline src/java.base/share/classes/java/lang/FdLibm.java:657]:
```suggestion
                    s1 = t1 + r;
```

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 8]
```diff
@@ -504,6 +513,449 @@ static double compute(double y, double x) {
...
+
+            // generate sqrt(x) bit by bit
+            ix0 += ix0 + ((ix1 & sign) >>> 31); // unsigned shift
+            ix1 += ix1;
+            q = q1 = s0 = s1 = 0;   // [q,q1] = sqrt(x)
+            r = 0x0020_0000;        // r = moving bit from right to left
+
+            while(r != 0) {
+                t = s0 + r;
+                if(t <= ix0) {
```
Comment (turbanoff) [inline src/java.base/share/classes/java/lang/FdLibm.java:641]:
```suggestion
                if (t <= ix0) {
```

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 9]
```diff
@@ -504,6 +513,449 @@ static double compute(double y, double x) {
...
+            double z = 0.0;
+            int sign = 0x8000_0000;
+            /*unsigned*/ int r, t1, s1, ix1, q1;
+            int ix0, s0, q, m, t, i;
+
+            ix0 = __HI(x);  // high word of x
+            ix1 = __LO(x);  // low word of x
+
+            // take care of Inf and NaN
+            if((ix0 & 0x7ff0_0000) == 0x7ff0_0000) {
```
Comment (turbanoff) [inline src/java.base/share/classes/java/lang/FdLibm.java:600]:
```suggestion
            if ((ix0 & 0x7ff0_0000) == 0x7ff0_0000) {
```

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 10]
```diff
@@ -504,6 +513,449 @@ static double compute(double y, double x) {
...
+                ix0 += ix0 + ((ix1 & sign) >>> 31); // unsigned shift
+                ix1 += ix1;
+                r >>>= 1; // unsigned shift
+            }
+
+            r = sign;
+            while (r != 0) {
+                t1 = s1 + r;
+                t  = s0;
+                if((t < ix0) ||
```
Comment (turbanoff) [inline src/java.base/share/classes/java/lang/FdLibm.java:655]:
```suggestion
                if ((t < ix0) ||
```

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 11]
```diff
@@ -504,6 +513,449 @@ static double compute(double y, double x) {
...
+            }
+            m >>= 1;        // m = [m/2]
+
+            // generate sqrt(x) bit by bit
+            ix0 += ix0 + ((ix1 & sign) >>> 31); // unsigned shift
+            ix1 += ix1;
+            q = q1 = s0 = s1 = 0;   // [q,q1] = sqrt(x)
+            r = 0x0020_0000;        // r = moving bit from right to left
+
+            while(r != 0) {
```
Comment (turbanoff) [inline src/java.base/share/classes/java/lang/FdLibm.java:639]:
```suggestion
            while (r != 0) {
```

File: src/java.base/share/classes/java/lang/StrictMath.java (hunk 1 above)
Comment (jddarcy) [reply src/java.base/share/classes/java/lang/StrictMath.java:312]: > I'll revert commenting out the intrinsic annotation before pushing.

Turns out some VM updates are needed to preserve the intrinsic if this method declaration is changed; I'll be contacting the HotSpot team for assistance.

File: src/java.base/share/classes/java/lang/FdLibm.java (hunk 11 above)
Comment (jddarcy) [reply src/java.base/share/classes/java/lang/FdLibm.java:639]: Thanks for the careful reading; will fix.

**Comment by jddarcy** [comment]
Adding hotspot-compiler to get get explicit review of small changes to enable intrinsics support for  StrictMath.sqrt after it was changed from a native method to a Java one. Specific VM changes:

```
diff --git a/src/hotspot/share/classfile/vmIntrinsics.hpp b/src/hotspot/share/classfile/vmIntrinsics.hpp
index 02be97362c7dc..6d1ad75ae688c 100644
--- a/src/hotspot/share/classfile/vmIntrinsics.hpp
+++ b/src/hotspot/share/classfile/vmIntrinsics.hpp
@@ -202,8 +202,8 @@ class methodHandle;
   do_intrinsic(_maxF_strict,              java_lang_StrictMath,   max_name,           float2_float_signature,    F_S)   \
   do_intrinsic(_minD_strict,              java_lang_StrictMath,   min_name,           double2_double_signature,  F_S)   \
   do_intrinsic(_maxD_strict,              java_lang_StrictMath,   max_name,           double2_double_signature,  F_S)   \
-  /* Special flavor of dsqrt intrinsic to handle the "native" method in StrictMath. Otherwise the same as in Math. */   \
-  do_intrinsic(_dsqrt_strict,             java_lang_StrictMath,   sqrt_name,          double_double_signature,   F_SN)  \
+  /* Additional dsqrt intrinsic to directly handle the sqrt method in StrictMath. Otherwise the same as in Math. */     \
+  do_intrinsic(_dsqrt_strict,             java_lang_StrictMath,   sqrt_name,          double_double_signature,   F_S)   \
                                                                                                                         \
   do_intrinsic(_floatIsInfinite,          java_lang_Float,        isInfinite_name,    float_bool_signature,      F_S)   \
    do_name(     isInfinite_name,                                  "isInfinite")                                         \
diff --git a/src/hotspot/share/interpreter/abstractInterpreter.cpp b/src/hotspot/share/interpreter/abstractInterpreter.cpp
index 875211d6d0527..8d6ef06754cfe 100644
--- a/src/hotspot/share/interpreter/abstractInterpreter.cpp
+++ b/src/hotspot/share/interpreter/abstractInterpreter.cpp
@@ -148,7 +148,7 @@ AbstractInterpreter::MethodKind AbstractInterpreter::method_kind(const methodHan
       case vmIntrinsics::_fmaD:              return java_lang_math_fmaD;
       case vmIntrinsics::_fmaF:              return java_lang_math_fmaF;
       case vmIntrinsics::_dsqrt:             return java_lang_math_sqrt;
-      case vmIntrinsics::_dsqrt_strict:      return native;
+      case vmIntrinsics::_dsqrt_strict:      return java_lang_math_sqrt;
       case vmIntrinsics::_Reference_get:     return java_lang_ref_reference_get;
       case vmIntrinsics::_Object_init:
         if (RegisterFinalizersAtInit && m->code_size() == 1) {
```

**Review by TobiHartmann — APPROVED** [review]
Hotspot changes look good to me.

**Review by turbanoff — APPROVED** [review]

**Pull request closed without merging**